### PR TITLE
DEX-465 Official Docker Image

### PR DESCRIPTION
**Pull request by DenisSerebryansky**
Main:
 * Images wavesplatform/matcher-node (Waves Node with Extension) and wavesplatform/matcher-server (DEX) introduced;
 * matcher-server is built upon openjdk:8-jre-slim-buster;
 * waves-integration-it image is built upon matcher-node;
 * dex-it image is built upon matcher-server;
 * ItCachedArtifactsPlugin introduced. It can be used for caching files that are pre-required in docker images;
 * ImageVersionPlugin introduced. It makes image names and complex tags;
 * CleanupDanglingImagesPlugin introduced. It removes all dangling images after building;
 * Draft of the README.md updates added (README_DOCKER.md)

Refactoring:
 * WavesNodeArtifactsPlugin slightly refactored;
 * Profiling moved to dex-it/docker;
 * start scripts for dex/dex-it/waves-ext/waves-integration-it renamed: start-matcher-server.sh/start-matcher-server-it.sh/start-matcher-node.sh/start-matcher-node-it.sh correspondingly;
 * WavesExtensionDockerPlugin, DexDockerPlugin and DexItDockerPlugin removed;
 * Removed few directories which became unnecessary for images building

### CODE REVIEW

**Comment by DenisSerebryansky** [comment]
Try it out!

1. `sudo chmod -R 777 /docker/waves`;
2. `sudo chmod -R 777 /docker/waves-dex`;
3. Put [local_waves.txt](https://github.com/wavesplatform/matcher/files/4913775/local_waves.txt) to `/docker/waves/waves-config/local.conf` (distribution config for the custom network);
4. Start wavesplatform/matcher-node: 
```
docker run -v /docker/waves/waves-data:/var/lib/waves -v /docker/waves/waves-config:/etc/waves -p 6871:6871 -e JAVA_OPTS="-Dwaves.rest-api.enable=yes -Dwaves.rest-api.bind-address=0.0.0.0 -Dwaves.wallet.password=myWalletSuperPassword" -e WAVES_NETWORK=stagenet --name matcher-node -ti wavesplatform/matcher-node
```
You will see that Node with DEX Extension started and works

5. Put [local_dex.txt](https://github.com/wavesplatform/matcher/files/4913799/local_dex.txt) to `/docker/waves-dex/runtime/local.conf` (gRPC target and account-storage settings);
6. Put [logback.txt](https://github.com/wavesplatform/matcher/files/4913806/logback.txt) to `/docker/waves-dex/runtime/logback.xml` (log management);
7. Start wavesplatform/matcher-server:
```
docker run -v /docker/waves-dex:/var/lib/waves-dex --link matcher-node:matcher-node -p 6886:6886 wavesplatform/matcher-server:latest
``` 

You will see that DEX started and works. Congratulations!

File: dex/build.sbt [hunk 1]
```diff
@@ -54,16 +57,46 @@ inConfig(Compile)(
...
+      val entryPointSh = s"$sourcesPath/bin/start-matcher-server.sh"
+
+      from("openjdk:8-jre-slim-buster")
+
+      runRaw(s"""mkdir -p $userPath $sourcesPath $userPath/runtime && \\
+                  |groupadd -g $groupId $group && \\
+                  |useradd -d $userPath -g $groupId -u $userId -s /bin/bash -M $user && \\
+                  |chown -R $userId:$groupId $userPath $sourcesPath && \\
+                  |chmod -R 755 $userPath $sourcesPath && \\
+                  |ln -fs $sourcesPath/log/ var/log/waves-dex""".stripMargin)
```
Comment (vsuharnikov) [inline dex/build.sbt:84]: Can we put this to `/var/lib/waves` instead? So the user will see file logs.

File: dex/README_DOCKER.md [hunk 2]
```diff
@@ -0,0 +1,90 @@
+## 7. Matcher Server & Matcher Node in Docker
```
Comment (vsuharnikov) [inline dex/README_DOCKER.md:1]: How a user can change JVM settings (e.g. the max heap size?)

File: dex/src/main/container/dex.conf [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+    # Bind address
+    address = "0.0.0.0"
+    # Bind port
+    port = 6886
+  }
+  # Default path for account.dat file
+  account-storage.encrypted-file.path = ${waves.dex.root-directory}"/account.dat"
+}
+
+include "/var/lib/waves-dex/runtime/local.conf"
```
Comment (vsuharnikov) [inline dex/src/main/container/dex.conf:15]: newline?

File: dex/src/main/container/dex.conf [hunk 4]
```diff
@@ -0,0 +1,15 @@
...
+  address-scheme-character = "W"
+  # REST API is available from the outside
+  rest-api {
+    # Bind address
+    address = "0.0.0.0"
+    # Bind port
+    port = 6886
+  }
+  # Default path for account.dat file
+  account-storage.encrypted-file.path = ${waves.dex.root-directory}"/account.dat"
```
Comment (vsuharnikov) [inline dex/src/main/container/dex.conf:12]: This is what I said above :)

File: waves-ext/src/main/resources/application.conf [hunk 5]
```diff
@@ -1,11 +1,10 @@
-# To enable this extension on NODE, add this line:
-# waves.extensions += "com.wavesplatform.dex.grpc.integration.DEXExtension"
+waves.extensions += "com.wavesplatform.dex.grpc.integration.DEXExtension"
 
 waves.dex {
   # gRPC integration settings for Waves Node
   grpc.integration {
     # Extension's host
-    host = localhost
+    host = "0.0.0.0"
```
Comment (vsuharnikov) [inline waves-ext/src/main/resources/application.conf:7]: It's better to do this only for a docker image. Probably this way: https://github.com/lightbend/config#optional-system-or-env-variable-overrides
> By setting the JVM property -Dconfig.override_with_env_vars=true it is possible to override any configuration value using environment variables even if an explicit substitution is not specified.

File: waves-ext/src/main/container/start-matcher-node.sh [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+echo "WAVES_WALLET_PASSWORD='${WAVES_WALLET_PASSWORD}'" | tee -a /var/log/waves/waves.log
+echo "JAVA_OPTS='${JAVA_OPTS}'" | tee -a /var/log/waves/waves.log
+
+java -Dlogback.stdout.level=${WAVES_LOG_LEVEL} \
+	-XX:+ExitOnOutOfMemoryError \
+	-Xmx${WAVES_HEAP_SIZE} \
+	-Dlogback.file.directory=/var/log/waves \
+	-Dconfig.override_with_env_vars=true \
+	${JAVA_OPTS} \
+	-cp "/opt/waves/lib/*:/usr/share/waves/lib/waves-all-${WAVES_VERSION}.jar" com.wavesplatform.Application /etc/waves/waves.conf
```
Comment (vsuharnikov) [inline waves-ext/src/main/container/start-matcher-node.sh:45]: newline?

File: waves-ext/src/main/container/start-matcher-node.sh [hunk 7]
```diff
@@ -0,0 +1,45 @@
+#!/bin/bash
+NETWORKS="mainnet testnet stagenet"
```
Comment (vsuharnikov) [inline waves-ext/src/main/container/start-matcher-node.sh:2]: Let's add a comment after shebang that we need this file until NODE-2172 is done

File: dex/README_DOCKER.md [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+
+## About the Images
+
+Matcher's team provides you with 2 docker images:
+  1. matcher-node;
```
[the remark this answers is not in the document]
Comment (DenisSerebryansky) [reply dex/README_DOCKER.md]: Fixed

File: dex/README_DOCKER.md [hunk 9]
```diff
@@ -0,0 +1,90 @@
...
+  2. matcher-server
+
+These images are focused on fast and convenient deployment of Waves Node with Matcher Extension and Matcher Server.
+
+First one is the Waves Node with pre-installed Matcher Extension (Matcher Node). It is based on the official [wavesnode](https://hub.docker.com/r/wavesplatform/wavesnode) image, so [it's configurations](https://github.com/wavesplatform/Waves/tree/master/docker#configuration-options) can be applied for the matcher-node image as well. If you're going to use volumes, please, pay attention to [data management](https://github.com/wavesplatform/Waves/tree/master/docker#managing-data) section.
+
+Matcher Extension in matcher-node image is already configured. Extension uses port 6887 for communication with Matcher Server via gRPC protocol. You can override Matcher Extension [default settings](https://github.com/wavesplatform/matcher/blob/master/waves-ext/src/main/resources/application.conf) in local.conf, which is located in `/docker/waves/waves-conf/` on your host machine (if you are using volume):
```
[the remark this answers is not in the document]
Comment (DenisSerebryansky) [reply dex/README_DOCKER.md]: Now everything is /matcher-node and /matcher-server

File: dex/README_DOCKER.md [hunk 10]
```diff
@@ -0,0 +1,90 @@
...
+Successful Node-Matcher interaction requires that they can see each other. Therefore, you should use container linking or a private network.
+The order in which containers are started matters! Matcher Node should be started first, then Matcher Server. 
+
+We will use container linking. Although this is the legacy practice, it is a bit simpler (you don't need to set up network before containers running).
+
+Also, Matcher Node and Matcher Server can be connected via network. In this case use network alias for the Matcher Node container and `waves.dex.waves-blockchain-client.grpc.target` setting. 
+
+#### Running Matcher Node
+```
+docker run -v /docker/waves/waves-data:/var/lib/waves -v /docker/waves/waves-config:/etc/waves -p 6869:6869 -p 6862:6862 -e JAVA_OPTS="-Dwaves.rest-api.enable=yes -Dwaves.rest-api.bind-address=0.0.0.0 -Dwaves.wallet.password=myWalletSuperPassword" -e WAVES_NETWORK=stagenet --name matcher-node -ti wavesplatform/matcher-node
```
[the remark this answers is not in the document]
Comment (DenisSerebryansky) [reply dex/README_DOCKER.md]: Done

File: dex/README_DOCKER.md [hunk 11]
```diff
@@ -0,0 +1,90 @@
...
+```
+
+Worth to note here that we specified container name by means of `--name matcher-node`. This name will be used as a link to Matcher Node during start of Matcher Server and also in Matcher Server setting `waves.dex.waves-blockchain-client.grpc.target`.
+
+#### Running Matcher Server
+```
+docker run -v /docker/waves-dex:/var/lib/waves-dex --link matcher-node:matcher-node -p 6886:6886 wavesplatform/matcher-server:latest
```
[the remark this answers is not in the document]
Comment (DenisSerebryansky) [reply dex/README_DOCKER.md]: Done

File: dex/README_DOCKER.md [hunk 12]
```diff
@@ -0,0 +1,90 @@
...
+```
+
+Here we specified link to Matcher Node container `--link matcher-node:matcher-node` and the mapping between directory on your host machine `/docker/waves-dex/` and inner container directory `/var/lib/waves-dex`. 
+
+Directory `/docker/waves-dex/data` will be used for storing Matcher data. You can put your own Matcher state here.
```
[the remark this answers is not in the document]
Comment (DenisSerebryansky) [reply dex/README_DOCKER.md]: Done

File: dex/README_DOCKER.md [hunk 13]
```diff
@@ -0,0 +1,90 @@
...
+
+If you are using encrypted file as a Matcher account storage, please, put `account.dat` file in this directory and do not forget to specify appropriate `waves.dex.account-storage` settings in `local.conf`.
```
[the remark this answers is not in the document]
Comment (DenisSerebryansky) [reply dex/README_DOCKER.md]: Mention dedicated setting `waves.dex.account-storage.type = "encrypted-file"`

File: dex/src/main/container/dex.conf (hunk 3 above)
Comment (DenisSerebryansky) [reply dex/src/main/container/dex.conf]: Done

File: waves-ext/src/main/container/start-matcher-node.sh (hunk 6 above)
Comment (DenisSerebryansky) [reply waves-ext/src/main/container/start-matcher-node.sh]: Done

File: waves-ext/src/main/container/start-matcher-node.sh (hunk 7 above)
Comment (DenisSerebryansky) [reply waves-ext/src/main/container/start-matcher-node.sh:2]: Done

File: waves-ext/src/main/resources/application.conf (hunk 5 above)
Comment (DenisSerebryansky) [reply waves-ext/src/main/resources/application.conf]: Done

File: dex/build.sbt (hunk 1 above)
Comment (DenisSerebryansky) [reply dex/build.sbt]: Done

File: dex/README_DOCKER.md (hunk 2 above)
Comment (DenisSerebryansky) [reply dex/README_DOCKER.md:1]: Done

**Pull request merged**